File: bindings/go-open/dhruv/tara.go
package dhruv

import (
	"runtime"

	"ctara-dhruv-core/bindings/go-open/internal/cabi"
)

type TaraCatalog struct {
	h cabi.TaraCatalogHandle
}

func LoadTaraCatalog(path string) (*TaraCatalog, error) {
	h, st := cabi.LoadTaraCatalog(path)
	if err := statusErr("tara_catalog_load", st); err != nil {
		return nil, err
	}
	c := &TaraCatalog{h: h}
	runtime.SetFinalizer(c, (*TaraCatalog).Close)
	return c, nil
}

func (c *TaraCatalog) Close() {
	if c == nil {
		return
	}
	runtime.SetFinalizer(c, nil)
	c.h.Free()
}

func (c *TaraCatalog) Compute(req TaraComputeRequest) (TaraComputeResult, error) {
	out, st := cabi.TaraComputeEx(c.h, req)
	return out, statusErr("tara_compute_ex", st)
}

func (c *TaraCatalog) GalacticCenterEcliptic(jdTdb float64) (SphericalCoords, error) {
	out, st := cabi.TaraGalacticCenterEcliptic(c.h, jdTdb)
	return out, statusErr("tara_galactic_center_ecliptic", st)
}

func TaraPropagatePosition(raDeg, decDeg, parallaxMas, pmRaMasYr, pmDecMasYr, rvKmS, dtYears float64) (EquatorialPosition, error) {
	out, st := cabi.TaraPropagatePosition(raDeg, decDeg, parallaxMas, pmRaMasYr, pmDecMasYr, rvKmS, dtYears)
	return out, statusErr("tara_propagate_position", st)
}

func TaraApplyAberration(direction [3]float64, earthVelAUDay [3]float64) ([3]float64, error) {
	out, st := cabi.TaraApplyAberration(direction, earthVelAUDay)
	return out, statusErr("tara_apply_aberration", st)
}

func TaraApplyLightDeflection(direction [3]float64, sunToObserver [3]float64, sunObserverDistanceAU float64) ([3]float64, error) {
	out, st := cabi.TaraApplyLightDeflection(direction, sunToObserver, sunObserverDistanceAU)
	return out, statusErr("tara_apply_light_deflection", st)
}

func TaraGalacticAnticenterICRS() ([3]float64, error) {
	out, st := cabi.TaraGalacticAnticenterICRS()
	return out, statusErr("tara_galactic_anticenter_icrs", st)
}

File: bindings/node-open/src/tara.js
'use strict';

const { addon } = require('./native');
const { checkStatus } = require('./errors');

class TaraCatalog {
  constructor(handle) {
    this._handle = handle;
    this._closed = false;
  }

  static load(path) {
    const r = addon.taraCatalogLoad(path);
    checkStatus('tara_catalog_load', r.status);
    return new TaraCatalog(r.handle);
  }

  close() {
    if (this._closed) return;
    addon.taraCatalogFree(this._handle);
    this._closed = true;
    this._handle = null;
  }

  galacticCenterEcliptic(jdTdb) {
    const r = addon.taraGalacticCenterEcliptic(this._handle, jdTdb);
    checkStatus('tara_galactic_center_ecliptic', r.status);
    return r.coords;
  }

  compute(request) {
    const r = addon.taraComputeEx(this._handle, request);
    checkStatus('tara_compute_ex', r.status);
    return r.result;
  }
}

function propagatePosition(raDeg, decDeg, parallaxMas, pmRaMasYr, pmDecMasYr, rvKmS, dtYears) {
  const r = addon.taraPropagatePosition(raDeg, decDeg, parallaxMas, pmRaMasYr, pmDecMasYr, rvKmS, dtYears);
  checkStatus('tara_propagate_position', r.status);
  return r.position;
}

function applyAberration(direction, earthVelAuDay) {
  const r = addon.taraApplyAberration(direction, earthVelAuDay);
  checkStatus('tara_apply_aberration', r.status);
  return r.direction;
}

function applyLightDeflection(direction, sunToObserver, sunObserverDistanceAu) {
  const r = addon.taraApplyLightDeflection(direction, sunToObserver, sunObserverDistanceAu);
  checkStatus('tara_apply_light_deflection', r.status);
  return r.direction;
}

function galacticAnticenterIcrs() {
  const r = addon.taraGalacticAnticenterIcrs();
  checkStatus('tara_galactic_anticenter_icrs', r.status);
  return r.direction;
}

module.exports = {
  TaraCatalog,
  applyAberration,
  applyLightDeflection,
  galacticAnticenterIcrs,
  propagatePosition,
};
